File: 0224-basic-calculator/0224-basic-calculator.cpp
class Solution {
public:
    int parse(const string& s, int& i){
        int n = (int)s.size();
        long long ans = 0;
        int sign = 1;
        while (i < n){
            if (s[i] == ' ')
                i++;
            else if (s[i] == '+'){
                sign = 1;
                i++;
            } else if (s[i] == '-'){
                sign = -1;
                i++;
            } else if (isdigit(s[i])){
                long long num = 0;
                while (i < n && isdigit(s[i])){
                    num = num * 10 + (s[i]-'0');
                    i++;
                }
                ans += sign * num;
                sign = 1;
            } else if (s[i] == '('){
                i++;
                int value = parse(s,i);
                ans += sign * value;
                sign = 1;
            } else { // s[i] == ')'
                i++;
                return ans;
            }
        }
        return ans;
    }
    int calculate(string s) {
        int i = 0;
        return parse(s,i);
    }
};

File: 0224-basic-calculator/0224-basic-calculator.py
class Solution:
    def calculate(self, s: str) -> int:
        i = 0
        n = len(s)
        def parse():
            nonlocal i
            sign = 1
            ans  = 0
            while i < n:
                if s[i] == ' ':
                    i += 1
                elif s[i] == '+':
                    sign = 1
                    i += 1
                elif s[i] == '-':
                    sign = -1
                    i += 1
                elif s[i].isdigit():
                    num = 0
                    while i < n and s[i].isdigit():
                        num = num * 10 + int(s[i])
                        i += 1
                    ans += sign * num
                    sign = 1
                elif s[i] == '(':
                    i += 1
                    value = parse()
                    ans += sign * value
                    sign = 1
                elif s[i] == ')':
                    i += 1
                    return ans
            return ans
        return parse()